File: adv19-1.php
<?php
class Def
{
    protected $i = 0;
    protected $cache = [];

    function readInput($filePath)
    {
        $lines = file($filePath);
        $patterns = explode(',', array_shift($lines));
        array_walk($patterns, function (&$val) { $val=trim($val); });
        array_shift($lines);
        $designs = $lines;
        array_walk($designs, function (&$val) { $val=trim($val); return true;});
        rslog($designs, '$designs');
        return [$patterns, array_filter($designs)];
    }

    function isDesignPossible($design, $patterns)
    {
        $finalCandidates = [];
        $curCandidates = [];
        $designArr = str_split($design);

        for ($i = 0; $i<count($designArr); $i++) {
            $curSymb = $designArr[$i];
            $curSymbHasAnyCandidates = false;
//            rslog($i, '$i processing........' . $curSymb);

            // Check old current candidates
            foreach ($curCandidates as $key => $cand) {
                list($candArr, $candStart) = $cand;
                $candI = $i - $candStart;
                if ($candArr[$candI] == $curSymb) {
                    $curSymbHasAnyCandidates = true;
                    if (count($candArr) == $candI+1) {
                        // End of candidate and it is valid
                        $finalCandidates[$candStart][] = [$candStart, count($candArr), $candArr];
//                        rslog(join($candArr), '"moving to final candidate"');
                        unset($curCandidates[$key]);
                    }
                } else {
                    // New symbol does not fit
                    unset($curCandidates[$key]);
                }
            }

            // Check patterns for a good start
            foreach ($patterns as $pattern) {
                $patternArr = str_split($pattern);
//                rslog($pattern[0] == $curSymb, "$pattern[0] == $curSymb");
                if ($patternArr[0] == $curSymb) {
                    // First letter of candidate matches
                    if (count($patternArr) == 1) {
                        $curSymbHasAnyCandidates = true;
                        // Its a one synbol pattern - it is a successfull candidate already
                        $finalCandidates[$i][] = [$i, 1, $patternArr];
//                        rslog(join($patternArr), '"new final candidate"');
                    } else {
//                        rslog($i, '$i');
//                        rslog(count($patternArr), 'count($patternArr)');
                        // its a few letters candidate which is just started - we will add it to candidates
                        if ($i < count($designArr)-1) {
//                            rslog(join($patternArr), '"adding candidate"');
                            $curSymbHasAnyCandidates = true;
                            // there will be more symbols to check in the design
                            $curCandidates[] = [$patternArr, $i];
                        }
                    }
                }
            }
            if (!$curSymbHasAnyCandidates) {
                // This symbol has no candidates coverage - no need to check further
                rslog($design, '$design <===============   FAILED');
                rslog($patterns, '$patterns');
                rslog($finalCandidates, '$finalCandidates');
                rslog($curCandidates, '$curCandidates');
                rslog($i, '$i < failed');
                return 0;
            }
        }


        rslog($design, '$design');
        rslog($finalCandidates, '$finalCandidates');


        $this->cache = [];
        return $this->checkPaths(0, $designArr, $finalCandidates);

//        rslog($design, '$design <============');
//        rslog($patterns, '$patterns');
//        rslog($finalCandidates, '$finalCandidates');
//        die();
    }

    function checkPaths($idx, &$designArr, &$finalCandidates, $i = 0)
    {
        if (isset($this->cache[$idx])) {
            return $this->cache[$idx];
        }
        $sum = 0;
        $i++;
        if (isset($finalCandidates[$idx]))
        foreach ($finalCandidates[$idx] as $cand) {
            //[$candStart, count($candArr), $candArr];
            $candLen = $cand[1];
            if ($candLen + $idx == count($designArr)) {
                // with this candidate we will reach the end of the design
                $sum++;
            } else {
                $sum += $this->checkPaths($idx + $candLen, $designArr, $finalCandidates, $i);
            }
        }
        $this->cache[$idx] = $sum;
        return $sum;
    }


    function run()
    {
        // Main Execution
//        $filePath = "data/19-1.example";
//        $filePath = "data/19-1.example2";
        $filePath = "data/19-1.inp";
        [$patterns, $designs] = $this->readInput($filePath);
        $foundCnt = 0;
        foreach ($designs as $i => $design) {
            $this->i = 0;
            rslog($design, '$design <===================== ' . $i);
            if ($cnt = $this->isDesignPossible($design, $patterns)) {
                $foundCnt += $cnt;
            }
        }
        rslog($foundCnt, '$foundCnt');
    }
}

File: adv19-1.py
class Def:
    def read_input(self, file_path):
        with open(file_path, 'r') as file:
            lines = file.readlines()

        patterns = [pattern.strip() for pattern in lines.pop(0).split(',')]
        lines.pop(0)
        designs = [line.strip() for line in lines]

        print(f"Designs: {designs}")
        return patterns, list(filter(None, designs))

    def is_design_possible(self, design, patterns):
        final_candidates = {}
        cur_candidates = []
        design_arr = list(design)

        for i, cur_symb in enumerate(design_arr):
            cur_symb_has_any_candidates = False

            # Check old current candidates (iterate in reverse to avoid index issues)
            for key in range(len(cur_candidates) - 1, -1, -1):
                cand_arr, cand_start = cur_candidates[key]
                cand_i = i - cand_start
                if cand_arr[cand_i] == cur_symb:
                    cur_symb_has_any_candidates = True
                    if len(cand_arr) == cand_i + 1:
                        final_candidates.setdefault(cand_start, []).append((cand_start, len(cand_arr), cand_arr))
                        cur_candidates.pop(key)
                else:
                    cur_candidates.pop(key)

            # Check patterns for a good start
            for pattern in patterns:
                pattern_arr = list(pattern)
                if pattern_arr[0] == cur_symb:
                    if len(pattern_arr) == 1:
                        cur_symb_has_any_candidates = True
                        final_candidates.setdefault(i, []).append((i, 1, pattern_arr))
                    elif i < len(design_arr) - 1:
                        cur_symb_has_any_candidates = True
                        cur_candidates.append((pattern_arr, i))

            if not cur_symb_has_any_candidates:
                print(f"Design {design} FAILED")
                print(f"Patterns: {patterns}")
                print(f"Final Candidates: {final_candidates}")
                print(f"Current Candidates: {cur_candidates}")
                return 0

        print(f"Design: {design}")
        print(f"Final Candidates: {final_candidates}")

        return self.check_paths(0, design_arr, final_candidates)

    def check_paths(self, idx, design_arr, final_candidates, depth=0):
        total = 0
        depth += 1

        if idx in final_candidates:
            for cand in final_candidates[idx]:
                cand_len = cand[1]
                if cand_len + idx == len(design_arr):
                    total += 1
                else:
                    total += self.check_paths(idx + cand_len, design_arr, final_candidates, depth)

        return total

    def run(self):
        file_path = "data/19-1.example2"
        patterns, designs = self.read_input(file_path)
        found_count = 0

        for i, design in enumerate(designs):
            print(f"Processing design {design} <===================== {i}")
            count = self.is_design_possible(design, patterns)
            if count:
                found_count += count

        print(f"Found Count: {found_count}")

# Usage Example
def_instance = Def()
def_instance.run()
